lib/Analysis/Allocation.cpp

```diff
@@ -88,30 +88,74 @@ getScratchConfigForCvtLayout(triton::gpu::ConvertLayoutOp op, unsigned &inVec,
   return paddedRepShape;
 }
 
-SmallVector<unsigned> getScratchConfigForReduce(triton::ReduceOp op) {
+/// ====================== reduce op config begin ==========================
+
+bool enableFastReduce(triton::ReduceOp op) {
+  auto srcTy = op.operand().getType().cast<RankedTensorType>();
+  auto srcLayout = srcTy.getEncoding();
+  return op.axis() == getOrder(srcLayout)[0];
+}
+
+SmallVector<unsigned> getScratchConfigForReduceBasic(triton::ReduceOp op) {
   auto srcTy = op.operand().getType().cast<RankedTensorType>();
   auto srcLayout = srcTy.getEncoding();
   auto srcShape = srcTy.getShape();
   auto axis = op.axis();
 
-  bool fastReduce = axis == getOrder(srcLayout)[0];
+  auto smemShape = convertType<unsigned>(srcShape);
+  smemShape[axis] = gpu::getThreadsPerWarp(srcLayout)[axis] *
+                    gpu::getWarpsPerCTA(srcLayout)[axis];
 
-  SmallVector<unsigned> smemShape;
-  for (auto d : srcShape)
-    smemShape.push_back(d);
+  return smemShape;
+}
+
+SmallVector<SmallVector<unsigned>>
+getScratchConfigsForReduceFast(triton::ReduceOp op) {
+  auto srcTy = op.operand().getType().cast<RankedTensorType>();
+  auto srcLayout = srcTy.getEncoding();
+  auto srcShape = srcTy.getShape();
+  auto axis = op.axis();
+
+  SmallVector<SmallVector<unsigned>> smemShapes(3);
+
+  /// shared memory block0
+  smemShapes[0] = convertType<unsigned>(srcShape);
+  smemShapes[0][axis] = gpu::getWarpsPerCTA(srcLayout)[axis];
 
-  if (fastReduce) {
-    unsigned sizeInterWarps = gpu::getWarpsPerCTA(srcLayout)[axis];
-    smemShape[axis] = sizeInterWarps;
+  /// shared memory block1
+  auto mod = op.getOperation()->getParentOfType<ModuleOp>();
+  unsigned numWarps = triton::gpu::TritonGPUDialect::getNumWarps(mod);
+  smemShapes[1].push_back(numWarps * 32);
```

Comment:
It seems that I made a mistake here. Although the read offset is `threadIdx.x`, only the first `warpsPerCTA` threads are actually working. The loads of other threads should be masked. Therefore we actually do not need three smemShapes and one should be enough. Let me think about this for more time.